src/toil/jobStores/aws/jobStore.py

```diff
@@ -407,141 +421,104 @@ def getEmptyFileStoreID(self, jobStoreID=None, cleanup=False, basename=None):
-                readable.seek(0)  # go to the 0th byte from the start of the file
-            except:
-                canDetermineSize = False
-            if canDetermineSize and fileSize > (5 * 1000 * 1000):  # only use multipart when file is above 5 mb
-                logger.debug("Uploading %s with size %s, will use multipart uploading", dstKey.name, fileSize)
-                chunkedFileUpload(readable=readable, bucket=dstKey.bucket, fileID=dstKey.name, file_size=fileSize)
-            else:
-                # we either don't know the size, or the size is small
-                logger.debug("Can not use multipart uploading for %s, uploading whole file at once", dstKey.name)
-                dstKey.set_contents_from_string(readable.read())
-        finally:
-            dstKey.bucket.connection.close()
+        dstObj = cls._getObjectForUrl(url)
+        resource = get_boto3_session().resource('s3')
+
+        logger.debug("Uploading %s", dstObj.key)
+        # uploadFile takes care of using multipart upload if the file is larger than partSize (default to 5MB)
+        uploadFile(readable=readable,
+                   resource=resource,
+                   bucketName=dstObj.bucket_name,
+                   fileID=dstObj.key,
+                   partSize=5 * 1000 * 1000)
 
     @staticmethod
-    def _getKeyForUrl(url, existing=None):
+    def _getObjectForUrl(url, existing: Optional[bool] = None):
         """
-        Extracts a key from a given s3:// URL. On return, but not on exceptions, this method
-        leaks an S3Connection object. The caller is responsible to close that by calling
-        key.bucket.connection.close().
+        Extracts a key (object) from a given s3:// URL.
 
         :param bool existing: If True, key is expected to exist. If False, key is expected not to
                exists and it will be created. If None, the key will be created if it doesn't exist.
 
-        :rtype: Key
+        :rtype: S3.Object
         """
         keyName = url.path[1:]
         bucketName = url.netloc
 
         botoargs = {}
         host = os.environ.get('TOIL_S3_HOST', None)
-        if host:
-            botoargs['host'] = host
         port = os.environ.get('TOIL_S3_PORT', None)
-        if port:
-            botoargs['port'] = int(port)
-        is_secure = os.environ.get('TOIL_S3_USE_SSL', True)
-        if is_secure == 'False':
-            is_secure = False
-            botoargs['is_secure'] = is_secure
+        protocol = 'https'
+        if os.environ.get('TOIL_S3_USE_SSL', True) == 'False':
+            protocol = 'http'
+        if host:
+            botoargs['endpoint_url'] = f'{protocol}://{host}' + f':{port}' if port else ''
 
-        if botoargs:
-            botoargs['calling_format'] = boto.s3.connection.OrdinaryCallingFormat()
+        # TODO: OrdinaryCallingFormat equivalent in boto3?
+        # if botoargs:
+        #     botoargs['calling_format'] = boto.s3.connection.OrdinaryCallingFormat()
```

Comment:
I wasn't able to find `calling_format` in boto3 and test the endpoint configuration added in #3370, but this should be properly tested.

Reply:
There isn't one I believe, so we can probably just drop it.  I'm fine with doing that in a subsequent PR and testing myself.  You'd need to download and install minion (which is kind of neat actually).  You're right, it needs proper testing, but that's for a future sprint I think.